src/addons/meta/runtime.c

```diff
@@ -0,0 +1,328 @@
+    ecs_vec_init_t(NULL, &rtt_ctx->moves, ecs_meta_rtt_call_data_t, 0);
+    ecs_vec_init_t(NULL, &rtt_ctx->copys, ecs_meta_rtt_call_data_t, 0);
+    hooks.lifecycle_ctx = rtt_ctx;
+    hooks.lifecycle_ctx_free = flecs_meta_rtt_free_lifecycle_ctx;
+
+    if (ctor) {
+      hooks.ctor = flecs_meta_rtt_ctor;
+    }
+    if (dtor) {
+      hooks.dtor = flecs_meta_rtt_dtor;
+    }
+    if (move) {
+      hooks.move = flecs_meta_rtt_move;
+    }
+    if (copy) {
+      hooks.copy = flecs_meta_rtt_copy;
+    }
+  } else {
+    hooks.lifecycle_ctx = NULL;
+  }
+  ecs_set_hooks_id(world, ti->component, &hooks);
+  return rtt_ctx;
+}
+
+// Checks if a struct member's types have hooks installed. If so, it generates and installs required hooks for the
+// struct type itself. These hooks will invoke the member hooks when necessary
+// returns true if hooks were actually configured, false otherwise (struct did not need hooks)
+static void flecs_meta_rtt_init_default_hooks_struct(ecs_world_t *world, ecs_entity_t component,
+                                                     const ecs_type_info_t *ti) {
+  // Obtain struct information to figure out what members it contains:
+  const EcsStruct *struct_info = ecs_get(world, component, EcsStruct);
+  ecs_assert(struct_info != NULL, ECS_INTERNAL_ERROR, NULL);
+
+  // These flags will be set to true if we determine we need to generate a hook of a particular type:
+  bool ctor_hook_required = false;
+  bool dtor_hook_required = false;
+  bool move_hook_required = false;
+  bool copy_hook_required = false;
+
+  // Iterate all struct members and see if hooks are required:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
+    ecs_member_t *m = ecs_vec_get_t(&struct_info->members, ecs_member_t, i);
+    const ecs_type_info_t *member_ti = ecs_get_type_info(world, m->type);
+    ctor_hook_required |= member_ti->hooks.ctor && member_ti->hooks.ctor != flecs_default_ctor;
+    dtor_hook_required |= member_ti->hooks.dtor != NULL;
+    move_hook_required |= member_ti->hooks.move != NULL;
+    copy_hook_required |= member_ti->hooks.copy != NULL;
+  }
+
+  // If any hook is required, then create a lifecycle context and configure a generic hook that will interpret that
+  // context:
+  ecs_meta_rtt_ctx_t *rtt_ctx = flecs_meta_rtt_configure_hooks(world, ti, ctor_hook_required, dtor_hook_required,
+                                                               move_hook_required, copy_hook_required);
+
+  if (!rtt_ctx)
+    return;  // no hooks required
+
+  // At least a hook was configured, therefore examine each struct member to build the vector of calls that will then
+  // be executed by the generic hook handler:
+  for (int i = 0; i < ecs_vec_count(&struct_info->members); i++) {
+    ecs_member_t *m = ecs_vec_get_t(&struct_info->members, ecs_member_t, i);
+    const ecs_type_info_t *member_ti = ecs_get_type_info(world, m->type);
+    if (ctor_hook_required) {
```

Comment:
That's fair, let's keep it this way for now then.

For ref, one way I've solved this in the past (different framework) is to compute the contiguous parts of a struct and memset/memcpy those:

```cpp
struct Foo {
  /* contiguous memset/memcpy-able block block */
  int foo;
  int bar;
  /* special member */
  char *str;
  /* contiguous memset/memcpy-able block block */
  int hello;
};
```
The code that generated the free routine would produce an array of instructions (similar to the array with callbacks you have, but with a bit more info) that had an instruction per contiguous block.

That's definitely more complex to implement, not suggesting we do that now.

Reply:
Yes, this sounds like what I was proposing, that is, to "defragment" the memsets. Instead of having "instructions",  `ecs_meta_rtt_call_data_t` is prepared for this: point the `hook` member to a pseudo-ctor that just memsets but ignores `type_info`. This pseudo_ctor memsets everything and uses `count` as size.

```c
void flecs_meta_rtt_pseudo_ctor(
    void *ptr, 
    int32_t count, 
    const ecs_type_info_t *ti)
{
    (void) type_info;
    ecs_os_memset(ptr, 0, count);
}
```
Then, configure `ecs_meta_rtt_call_data_t` instance to have in `count` the sum of all sizes to memset.
Note that this is sort of a "private" hook, part of the call list and thus it is not directly callable from outside; nobody else would be able to call it with a bad `count` or anything else. So it looks like a hack, but it simplifies `ecs_meta_rtt_call_data_t` and the generic hook that has to execute the call list: It won't have to interpret "instructions", just blindly execute the call list, which would be faster I think.

The same applies to move and copy hooks, and that is why I'm doing:
```c
move_data->count = m->size * m->count;
move_data->hook.move = flecs_meta_rtt_default_move;
```

`flecs_meta_rtt_default_move` being:

```c
static void flecs_meta_rtt_default_move(void *dst_ptr, void *src_ptr, int32_t count, const ecs_type_info_t *type_info) {
  (void) type_info;  // silence unused warning
  ecs_os_memcpy(dst_ptr, src_ptr, count);
}
```

A second pass (future implementation) would then merge all call datas with `flecs_meta_rtt_default_move` into a big one with `count` equal to the sum of all sizes.

Reply:
I'm not a fan of "abusing" `count` as a `size` member. An invariant of the hook interface is that the memory pointed to by `dst_ptr` and `src_ptr` is `count * type_info->size`.

In a future performance optimization that collapses multiple fields into a single memset/memcpy, you could simply leave the callback to `NULL`, and call memset/memcpy directly. This is also better performance, since you don't have the overhead of calling a function pointer.

Reply:
fixed